File: huffmanED/huffman.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<queue>

using namespace std;

unordered_map<string, string> encode;
unordered_map<string, string> decode;

class Node{

    public:

        string data = "";
        int freq = 0;
        Node* left = nullptr;
        Node* right = nullptr;

        Node(string data, int freq, Node* left, Node* right){
            this->data = data;
            this->freq = freq;
            this->left = left;
            this->right = right;
        }

        bool operator < (const Node& o)const{
            return this->freq > o.freq;
        }
};

class huffman{

    private:
        void huffmanTree(string str){
            
            priority_queue<Node*> pq; 
            vector<int> freqMap(26, 0);

            for(int i = 0; i < str.size(); i++){
                freqMap[str[i] - 'a']++;
            }

            for(int i = 0; i < freqMap.size(); i++){
                if(freqMap[i] > 0){
                    // cout<<char('a' + i)<< "; ";
                    pq.push(new Node((char)('a' + i) + "", freqMap[i], nullptr, nullptr));
                }
            }

            while(pq.size() > 1){
                Node* n1 = pq.top();
                pq.pop();
                Node* n2 = pq.top();
                pq.pop();

                pq.push(new Node(n1->data + n2->data, n1->freq + n2->freq, n1, n2));
            }

            Node* root = pq.top();
            cout<<root->data<<endl;
            huffmanTraversal(root, "");
        }

        void huffmanTraversal(Node* root, string binaryCode){

            if(root->left == nullptr && root->right == nullptr){
                encode[root->data] = binaryCode;
                decode[binaryCode] = root->data;

                // cout<<decode[binaryCode]<<" ";
                return;
            }

            huffmanTraversal(root->left, binaryCode + "0");
            huffmanTraversal(root->right, binaryCode + "1");
        }

    public:

        huffman(string str){
            huffmanTree(str);
        }

        string encodeString(string str){

            string encoded = "";
            for(int i = 0; i < str.size(); i++){
                encoded += encode[str[i] + ""];
            }
            return encoded;
        }

        string decodeString(string str){

            string decoded = "";
            int i = 0;

            for(int j = 1; j <= str.size(); j++){
                string s = str.substr(i, j);
                if(decode.find(s) != decode.end()){
                    decoded += decode[s];
                    i = j;
                }
            }
            return decoded;
        }
};

int main(){

    huffman hed("hskajdyekuqhowpqowouwytremnzxbcvasaldjsfhjowiqowqojpqsakja");

    // string en = hed.encodeString("harshit");
    // string de = hed.decodeString(en);

    // cout<<en<<" -> "<<de<<endl;

    // cout<<hed.encodeString("j");

    return 0;
}

File: huffmanED/huffman.java
package huffmanED;

import java.util.HashMap;
import java.util.PriorityQueue;

public class huffman{
    
    private HashMap<String, String> encode = new HashMap<>();
    private HashMap<String, String> decode = new HashMap<>();

    private class Node implements Comparable<Node>{
        String data = "";
        int freq = 0;
        Node left = null;
        Node right = null;
        
        public Node(String data, int freq, Node left, Node right){
            this.data = data;
            this.freq = freq;
            this.left = left;
            this.right = right;
        }

        @Override
        public int compareTo(Node o) {
            return this.freq - o.freq;
        }
    }

    public huffman(String str){
        huffmanTree(str);
    }

    public void huffmanTree(String str){

        PriorityQueue<Node> pq = new PriorityQueue<>();
        int[] freq = new int[26];
        
        for(int i = 0; i < str.length(); i++){
            freq[str.charAt(i) - 'a']++;
        }

        for(int i = 0; i < freq.length; i++){
            if(freq[i] > 0)
                pq.add(new Node((char)('a' + i) + "", freq[i], null, null));
        }

        while(pq.size() > 1){
            Node n1 = pq.poll();
            Node n2 = pq.poll();

            pq.add(new Node(n1.data + n2.data, n1.freq + n2.freq, n1, n2));
        }
        Node root = pq.poll();
        huffmanTraversal(root, "");
    }

    public void huffmanTraversal(Node root, String binaryCode){
        
        // if(root == null){
        //     return;
        // }
        
        if(root.left == null && root.right == null){
            encode.put(root.data, binaryCode);
            decode.put(binaryCode, root.data);
            return;
        }
    
        huffmanTraversal(root.left, binaryCode + '0');
        huffmanTraversal(root.right, binaryCode + '1');
    }

    public String encodeString(String str){
        
        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < str.length(); i++){
            sb.append(encode.get(str.charAt(i) + ""));
        }
        
        return sb.toString(); 
    }

    public String decodeString(String str){

        StringBuilder sb = new StringBuilder();
        int i = 0;
        for(int j = 1; j <= str.length(); j++){
            String s = str.substring(i, j);
            if(decode.containsKey(s) == true){
                sb.append(decode.get(s));
                i = j;
            }
        }

        return sb.toString();
    }
}